src/pyodide/internal/workers.py

```diff
@@ -149,54 +157,208 @@ def json(
 
     def __contains__(self, key: str) -> bool:
-        return self.js_form_data.has(key)
+        return self._js_form_data.has(key)
 
     def values(self) -> Generator[FormDataValue, None, None]:
-        yield from self.js_form_data.values()
+        for val in self._js_form_data.values():
+            yield _js_value_to_py(val)
 
     def keys(self) -> Generator[str, None, None]:
-        yield from self.js_form_data.keys()
+        yield from self._js_form_data.keys()
 
     def __iter__(self):
         yield from self.keys()
 
     def items(self) -> Generator[tuple[str, FormDataValue], None, None]:
-        for item in self.js_form_data.entries():
-            yield (item[0], item[1])
+        for k, v in self._js_form_data.entries():
+            yield (k, _js_value_to_py(v))
 
     def __delitem__(self, key: str):
         self.delete(key)
 
     def __len__(self):
         return len(self.keys())
+
+    def get_all(self, key: str) -> list[FormDataValue]:
+        return list(map(lambda x: _js_value_to_py(x), self._js_form_data.getAll(key)))
+
+    @property
+    def js_object(self) -> "js.FormData":
+        return self._js_form_data
+
+
+def _supports_buffer_protocol(o):
+    try:
+        # memoryview used only for testing type; 'with' releases the view instantly
+        with memoryview(o):
+            return True
+    except TypeError:
+        return False
+
+
+@contextmanager
+def _make_blob_entry(e):
+    if isinstance(e, str):
+        yield e
+        return
+    if isinstance(e, Blob):
+        yield e._js_blob
+        return
+    if hasattr(e, "constructor") and (e.constructor.name in ("Blob", "File")):
+        yield e
+        return
+    if _supports_buffer_protocol(e):
+        px = create_proxy(e)
+        buf = px.getBuffer()
+        try:
+            yield buf.data
+            return
+        finally:
+            buf.release()
+            px.destroy()
+    raise TypeError(f"Don't know how to handle {type(e)} for Blob()")
+
+
+BlobValue = (
+    "str | bytes | js.ArrayBuffer | js.TypedArray | js.DataView | js.Blob | Blob"
+)
+
+
+class BlobEnding(StrEnum):
+    TRANSPARENT = "transparent"
+    NATIVE = "native"
+
+
+class BlobKwargs(TypedDict, total=False):
+    content_type: str
+    endings: BlobEnding
+
+
+class Blob:
+    def __init__(
+        self,
+        blob_parts: "Iterable[BlobValue] | BlobValue",
+        **options: Unpack[BlobKwargs],
+    ):
+        if "content_type" in options:
+            options["type"] = options["content_type"]
+        if "endings" in options:
+            options["endings"] = str(options["endings"])
+
+        is_single_item = not hasattr(blob_parts, "__iter__")
+        if is_single_item:
+            if "type" not in options:
+                # Inherit the `type` if we have a single item.
+                self._js_blob = (
+                    blob_parts.js_object if isinstance(blob_parts, Blob) else blob_parts
+                )
+                return
```

Comment:
Won't this mess up if someone passed `endings`? I'd suggest:
```py
# Inherit the `type` if we have a single item.
if not content_type and isinstance(blob, Blob):
   content_type = blob.content_type
if not content_type and hasattr(blob, "constructor") and blob.constructor.name == "Blob":
   content_type = blob.type
```
and then always create a new blob in the same place.

Reply:
If I do this then I also have to handle `filename`, but yeah, I'll do that

Reply:
Well if you want to allow it, that sounds like a reasonable helpful behavior.